Parquet Writer

xiaodaigh opened the pull request.

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,529 @@
+        recommended_chunks = 1
+    end
+
+    @showprogress for (coli, colname_sym) in enumerate(colnames)
```
Would rather not add dependency on `ProgressMeter`, at least not at this stage.

xiaodaigh replied:
I will remove it. I think reporting progressing is very good for user experience. Why do you not want it at this stage?

xiaodaigh replied:
have removed

tanmaykm replied:
> Why do you not want it at this stage?

Progressbar is a ui, should be at a higher level. Lower level functions can give out metrics for the progressbar. We are just building out the API now, even where to plug metrics in is not clear yet IMO.

xiaodaigh replied:
Ok. So you want to provide a callback mechanism for UI stuff?

tanmaykm commented on README.md:
```diff
@@ -140,3 +142,28 @@ julia> for v in values
+## Writer
+
+You can write any Tables.jl column accessible tables that contains columns of these types and their union with `Missing`: `Int32`, `Int64`, `String`, `Bool`, `Float32`, `Float64`
+
```
Would be good to also document types not supported yet, e.g timestamps (int96/int128)

xiaodaigh replied:
added some documentation

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,529 @@
+    Bool => PAR2._Type.BOOLEAN,
+    Int32 => PAR2._Type.INT32,
+    Int64 => PAR2._Type.INT64,
+    #INT96 => 3,  // deprecated, only used by legacy implementations. # not supported
```
Judging from the interactions in this package, it seems to me that this is not too infrequent, especially for timestamps. Maybe Julia datetime should be mapped to this.

xiaodaigh replied:
Not sure if I understand this comment. In the next version, I can look into how dates can be stored.

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,529 @@
+    Float32 => PAR2._Type.FLOAT,
+    Float64 => PAR2._Type.DOUBLE,
+    String => PAR2._Type.BYTE_ARRAY, # BYTE_ARRAY
+    # FIXED_LEN_BYTE_ARRAY => 7,
```
maybe fixed length byte arrays can be supported too, as they are written out similar to how strings are written?

xiaodaigh replied:
We need type that maps to `Fixed_Len_Byte_array`, currently, we don't have any types that maps to that.

E.g. A vector of FIXED length arrays can map to that. We can open up the format so that other users can add to it.

xiaodaigh commented:
> Thanks for initiating this. This seems like a good start. Any idea why the CI has failed?
> I had a quick look and have added a few suggestions.
> Will be looking at it more in the next few days.

Seems to be failing due to `nonmissingtype` which is not available in Base prior to Julia 1.3. Added. Hope no further issues. `Missings.jl` is added to deps for this reason.

I added CategoricalyArrays.jl purely so that I can use it to tell users it's not supported, please let me know if you think it's ok.

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,528 @@
+    # do not support dictionary with more than 127 levels
+    # TODO relax this 127 restriction
+    if length(uvals) > 127
+        @warn "More than 127 levels in dictionary. This is not supported at this stage."
```
should probably throw an error?

xiaodaigh replied:
I think we shouldn't throw error here as it's just something no supported by Parquet.jl at this stage. It's actually supported by Parquet.

tanmaykm replied:
Warning may have been fine if the data were actually written, but suboptimally. For example, warning that there's no CRC being calculated now is probably fine, because CRC may be considered as an optional check.

But since here it skips writing some data altogether, there is a risk of the user trying to use the resulting incomplete file further down in their workflow. It is better to throw an error IMO.

xiaodaigh replied:
It doesn't write a corrupted file, if that's your concern. It is actually fine to skip the writing of the dictionary. In this first version, dictionary pages are not used at all. In the next version, we will use dictionary pages properly.

xiaodaigh replied:
@tanmaykm Are you OK to merge it as it for this version? If not I can remove the dictionary writing code altogether this round. And include it in next version without that 127 level restriction.

As I said above, currently the dictionary is not used in writing anyway, so it's fine as it is. But throwing an error is definitely not correct as the file doesn't need a dictionary.

tanmaykm replied:
Ah okay, so `write_col_dict` is not called anywhere? I always find it less confusing to comment out unused code.

xiaodaigh replied:
commented out in latest push (soon)

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,528 @@
+            # use the bit packing algorithm to write the
+            # definition_levels
+
+            bytes_needed = ceil(Int, length(colvals) / 8sizeof(UInt8))
```
> 8sizeof(UInt8)

is that correct?

xiaodaigh replied:
yes. as you can see in the test. all tests pass.

This write-only deals with un-nested columns so it packs the values such that `0101` represents where the value is (reading from right to left) non-missing, missing, non-missing, missing,  and the pattern `011` would read missing, missing, non-missing. And there are 8 bits in a byte. So `8sizeof(UInt8)` is the correct number of bits in a byte`. I could've written 8 but I think this is more clear in terms of the intention.

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,528 @@
+        data_to_compress_io = IOBuffer()
+
+        if Missing <: T
+            # if there is missing
```
better to separate out writing of definition levels into a separate function

xiaodaigh replied:
done

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,528 @@
+            uncompressed_size = uncompressed_page_size + size_of_page_header_defn_repn,
+            compressed_size = compressed_page_size + size_of_page_header_defn_repn,
+        )
+    elseif encoding == PAR2.Encoding.PLAIN_DICTIONARY
```
separate out methods for writing column chunks in different encodings.

xiaodaigh replied:
done

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,528 @@
+function write_col(fileio, colvals::AbstractArray{T}, colname, encoding, codec; num_chunks = 1) where T
+    """Write a column to a file"""
+    # TODO turn writing dictionary on
+    if false
```
comment out code instead of using `if false`

xiaodaigh replied:
`if false` makes it easier to remember that when turning it on you should remove the branch in `else`.

I have removed it.

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,528 @@
+    else
+        recommended_chunks = 1
+    end
+
```
Should split the writer method at this point into a lower level method that can take more details - like the schema, chunk sizes, encodings, codecs and such.

Heuristics that are needed to figure those out can then be outside in the higher level caller function.

The lower level method will be generic can be used to write other formats, for example one that can support nested structures.

xiaodaigh replied:
done. I made a `_write_parquet` method

xiaodaigh commented on src/writer.jl:
```diff
@@ -0,0 +1,528 @@
+    # do not support dictionary with more than 127 levels
+    # TODO relax this 127 restriction
+    if length(uvals) > 127
+        @warn "More than 127 levels in dictionary. This is not supported at this stage."
```
```suggestion
        @warn "More than 127 levels in dictionary. Parquet.jl does not support this at this stage."
```

xiaodaigh commented:
Done all the changes. Please review. IT's passing all tests locally on 1.0.5 and 1.4.1. The resultant file can be read by R and Python.

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,596 @@
+
+    position_before_page_header_write = position(fileio)
+    write_thrift(fileio, data_page_header)
+    size_of_page_header_defn_repn = position(fileio) - position_before_page_header_write
```
`write_thrift` should return the number of bytes written, why not use that directly instead of calling `position` on file?

xiaodaigh replied:
I see, that's a good idea. I didn't think of that, even though that's the normal thing to do. I will update.

xiaodaigh replied:
done

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,596 @@
+
+    write_thrift(fileio, filemetadata)
+
+    filemetadata_size = position(fileio) - position_before_filemetadata_write
```
use return value of `write_thrift` instead of `position`

xiaodaigh replied:
ok

xiaodaigh replied:
done

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,596 @@
+    )
+end
+
+function _write_parquet(tbl, path, nchunks; encoding::Dict{String, Int32}, codec::Dict{String, Int32})
```
This method should accept a schema too from the caller. And the data for each column as abstractvectors. We should not have a dependency on Tables here.

xiaodaigh replied:
> This method should accept a schema too from the caller

The schema is completely determined by the vector, so I propose just to pass the `AbstractVector`s and have the algorithm determine the right schema for each vector. What do you think?



> And the data for each column as abstractvectors.

Ok. That would work. I just make sure that no unnecessary allocations occur.



> We should not have a dependency on Tables here.

OK

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,596 @@
+    for val in colvals
+        # for string it needs to be stored as BYTE_ARRAY which needs the length
+        # to be the first 4 bytes UInt32
+        write(io, val |> sizeof |> UInt32)
```
don't we need to take care of endianness here?

xiaodaigh replied:
As shown in the test, the data can be read back correctly. The assumption is that we are using little-endian machines? Or we have to check that and write it in little-endian ways?

tanmaykm commented on src/writer.jl:
```diff
@@ -0,0 +1,596 @@
+
+    encoded_defn_data_length = length(bitpacking_header) + bytes_needed
+    # write the definition data
+    write(data_to_compress_io, UInt32(encoded_defn_data_length))
```
Don't we need to take care of endianness here?
I see this in few other places as well.

xiaodaigh replied:
> Don't we need to take care of endianness here?

What do you actually mean by that?

xiaodaigh replied:
I don't have experience writing code for non-little-endian machines, so I assume I am writing the data in little-endian way.

tanmaykm replied:
Maybe just `hton` will do it? But maybe this can be done later too.

xiaodaigh replied:
![image](https://user-images.githubusercontent.com/4497189/82625817-793f3880-9c29-11ea-8614-75172ca00ce6.png)

We should ensure the written data is in little-endian?

xiaodaigh replied:
![image](https://user-images.githubusercontent.com/4497189/82625950-daffa280-9c29-11ea-8cf1-4efc71f43d06.png)

sounds like `htol` is it. So wrap all writes in `htol`?

tanmaykm replied:
yes, for int32

xiaodaigh commented:
The remaining question is little-endian ness of the write. Since it passes all tests I'd say merge it and if it errors on non-little-endian machines, then ppl can raise a bug report? 

Since it's passing tests on travis-ci, it's useful as it is for majority of users.

tanmaykm commented:
Thanks for the update. I think it is not too difficult to take care of the endianness, will try an suggest something. I think this can be merged soon, but I would like to take just a bit more time to mull on what the end user API should be and do some changes if we should at this stage.

xiaodaigh commented:
Not sure why it's failing appveyor

xiaodaigh commented:
Resolved conflict with master

xiaodaigh commented:
Broke by #81 #82 

Tried to accommodate #81 but #82 is definitely a bug, as R and Python can read it OK.

xiaodaigh commented on test/test_writer.jl:
```diff
@@ -0,0 +1,59 @@
+
+
+    for (colnum, col_chunk) in enumerate(col_chunks)
+        println(colnum)
```
```suggestion

```

tanmaykm commented:
I feel we should have a write API that allows data to be written iteratively. Something like:

```
parwriter = ParFileWriter(path)

for chunk in table_chunks # chunk being a table itself
    write(parwriter, chunk)
end
```

Each iteration will add a row group to the file. And the file metadata can be updated at the end of the iteration. Looks like the underlying functions are already there here. So it should not be difficult. What do you think?

xiaodaigh commented:
> What do you think?

I am ok with that in addition to an API for writing a full table. Cos the `table_chunks` have to come from somewhere, so for ppl to use it they would need a way to chunk tables, which is rarely done. I'd say 90% of use cases are where the user just provides a Table so it's most straightforward to just write it.

How about this? We merge this one first, if no objection to the write API as it is, then in a future release we write this chunk based writer (to cover what I guess is just 10% of use cases).

xiaodaigh commented:
Moer serious issue is that tests broke by #82

xiaodaigh commented on src/Parquet.jl:
```diff
@@ -7,13 +7,20 @@ using CodecZstd
+    using Missings: nonmissingtype
+end
+
+const PARQUET_JL_VERSION = v"0.5.1"
```
```suggestion
const PARQUET_JL_VERSION = v"0.5.3"
```

xiaodaigh commented on test/test_writer.jl:
```diff
@@ -0,0 +1,70 @@
+    # the file is very small so only one rowgroup
+    col_chunks = columns(pf, 1)
+
+    colnum=12
```
```suggestion
```

xiaodaigh commented:
Merged latest changes from master and passing all tests. If there are no further changes requested from the writer, can I suggest we merge this. And implement the chunk based writer in the next release?

tanmaykm commented:
:+1: will merge in a bit

tanmaykm commented:
Oops, should have squashed and merged. Maybe I should rebase on the master, else it will become too unwieldy?

tanmaykm commented:
doesn't look possible now

xiaodaigh commented:
> Maybe I should rebase on the master, else it will become too unwieldy?

Feel free. To me it's not that big a deal.
